### ISSUE TITLE
[API Proposal]: Add System.Net.WebUtility.HtmlDecode(ReadOnlySpan<char> s)

### ISSUE THREAD

**Issue by lsoft** [dotnet/runtime#81411]
### Background and motivation

No way to decode XML-encoded string from `Span` exists now. We forced to allocate an intermediate string:

```csharp
var mystring = @"my string !@#$%^&amp;*()_+|-=\&#39;;[]{},./&lt;&gt;?".AsSpan();
string result = global::System.Net.WebUtility.HtmlDecode(mystring.ToString());
return result;
```

### API Proposal

```diff
public class HtmlDecode
{
+    public ReadOnlySpan<char> HtmlDecode(ReadOnlySpan<char> s);
      //OR
+    public string HtmlDecode(ReadOnlySpan<char> s);
}
```

### API Usage

```csharp
var mystring = @"my string !@#$%^&amp;*()_+|-=\&#39;;[]{},./&lt;&gt;?".AsSpan();
string result = global::System.Net.WebUtility.HtmlDecode(mystring.ToString());
return result;
```

### Risks

No risks.

**Comment by rzikm** [comment]
> +    public ReadOnlySpan<char> HtmlDecode(ReadOnlySpan<char> s);

How would this work? Would it allocate a new char array and return ROS to it?

If you want span-based API, then It would need to accept Span<char> and work in-place, or accept another span as destination.

**Comment by lsoft** [comment]
@rzikm yeah, I can easily miss here. But if the string no need to be decoded actually (for example `test123` has nothing to decode), then here probably no sense to allocate for result string. I guess we want to process such situation efficiently.

may be something like this:

```diff
public class HtmlDecode
{
+    public bool HasSomethingToHtmlDecode(ReadOnlySpan<char> s);
+    public string HtmlDecode(ReadOnlySpan<char> s);
}
```

from my POV, I will be happy with any API without intermediate string allocation.

**Comment by rzikm** [comment]
> I will be happy with any API without intermediate string allocation.

It seems to me that the other method (`string HtmlDecode(ReadOnlySpan<char> s)`) might be enough. You are still going to use the result as `string`.

If you would want to avoid allocating the return value as a string, then I suppose you would need something like

`public int HtmlDecode(ReadOnlySpan<char> source, Span<char> destination)`. I.e. decode `source` to `destination` and return the length of the result. In case the `destination` span is too short, only first `destination.Length` characters are stored.

Or, something similar, maybe we already use such a pattern elsewhere in the codebase.

**Comment by MihaZupan** [comment]
The pattern is usually
```c#
public static bool TryFoo(ReadOnlySpan<char> source, ReadOnlySpan<char> destination, out int charsWritten);
```

There's a very similar discussion on #40603.

**Comment by lsoft** [comment]
@rzikm `public int HtmlDecode(ReadOnlySpan<char> source, Span<char> destination)` assumes copying even if the string is not needed to be decoded. Not sure we want such pattern... anyway, maintainers will decide :)

**Comment by MihaZupan** [comment]
well, where are you getting the input from, and what are you doing with the decoded value?

**Comment by gfoidl** [comment]
In https://github.com/dotnet/runtime/issues/81411#issuecomment-1410311499
```diff
-public static bool TryFoo(ReadOnlySpan<char> source, ReadOnlySpan<char> destination, out int charsWritten);
+public static bool TryFoo(ReadOnlySpan<char> source, Span<char> destination, out int charsWritten);
```
I'll hide my comment after the copy pasta is fixed.

**Comment by antonfirsov** [comment]
Triage: sounds like a valuable idea, the exact API shape is yet to be figured out. Not critical for 8.0.

**Comment by MihaZupan** [comment]
Any thoughts on this https://github.com/dotnet/runtime/issues/81411#issuecomment-1410374784 @lsoft?

**Comment by lsoft** [comment]
@MihaZupan I think that, from API perspective, what I'm doing with the result is of little importance. API should not perform useless operations, and copying bytes from one place to another is not a subject of `HtmlDecode` API.

I tried to express my feelings that no allocation + no expesive ops should be performed if possible.

from a general point of view returns `Span` is more flexible than return `string`, because of `Span<char>` can be easily converted to `string`.

so I think the following API would be ideal:

```C#
public class HtmlDecode
{
+      public bool TryHtmlDecode(ReadOnlySpan<char> source, Span<char> destination, out int charsWritten);
}
```

that method shouldn't copy if `source` has nothing decode.

anyway, maintainers will weigth the options and will choose the optimal one.

**Comment by lsoft** [comment]
guys, I think the unmet needs here is visible fine. This need is the primary point I want to highlight. The specific method signature is secondary point. I'm fine with any decision that will you choose. Thanks! 👍
